### ISSUE TITLE
Synthetic source ignored source bug

### ISSUE THREAD

**Issue by martijnvg** [elastic/elasticsearch#113867]
Failures like this one are being reported:

```
Failed to execute phase [query], all shards failed; shardFailures {[JVfxnU8cQ96emScnvl2-Jw][.ds-logs-elastic_agent-default-2024.09.25-000001][0]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-7bf7c67dc9-285vg][10.2.83.215:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.search.fetch.FetchPhaseExecutionException: Fetch Failed [Error running fetch phase for doc [54935]]
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:80)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:191)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:81)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:754)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:699)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:559)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:79)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:76)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:101)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:992)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: com.fasterxml.jackson.core.JsonParseException: Duplicate field 'file'
 at [Source: (byte[])"{"file":{"line":647},"function":"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).watchRuntimeComponents","file":{"name":"coordinator/coordinator.go"}}"; line: 1, column: 155]
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.JsonReadContext._checkDup(JsonReadContext.java:250)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.JsonReadContext.setCurrentName(JsonReadContext.java:244)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:758)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.JsonGenerator._copyCurrentContents(JsonGenerator.java:2657)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.JsonGenerator.copyCurrentStructure(JsonGenerator.java:2638)
	at org.elasticsearch.xcontent.impl@9.0.0/org.elasticsearch.xcontent.provider.json.JsonXContentGenerator.copyCurrentStructure(JsonXContentGenerator.java:540)
	at org.elasticsearch.xcontent@9.0.0/org.elasticsearch.xcontent.XContentBuilder.copyCurrentStructure(XContentBuilder.java:1237)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper$TypeUtils.decodeAndWriteXContent(XContentDataHelper.java:603)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper$TypeUtils$11.decodeAndWrite(XContentDataHelper.java:502)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper.decodeAndWrite(XContentDataHelper.java:96)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper.writeMerged(XContentDataHelper.java:129)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$FieldWriter$IgnoredSource.writeTo(ObjectMapper.java:1084)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:1004)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$FieldWriter$FieldLoader.writeTo(ObjectMapper.java:1059)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:1004)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(SourceLoader.java:217)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:181)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$LeafWithMetrics.source(SourceLoader.java:146)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:290)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:233)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:162)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:72)
	... 15 more
	Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder
		at org.elasticsearch.xcontent@9.0.0/org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1282)
		at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:180)
		... 20 more
	Caused by: java.io.IOException: Unclosed object or array found
		at org.elasticsearch.xcontent.impl@9.0.0/org.elasticsearch.xcontent.provider.json.JsonXContentGenerator.close(JsonXContentGenerator.java:600)
		at org.elasticsearch.xcontent@9.0.0/org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1280)
		... 21 more
}
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:727)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:417)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:759)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:510)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:338)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:54)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:719)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1775)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1508)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1642)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1617)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:44)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:45)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:152)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:29)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:992)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: org.elasticsearch.ElasticsearchException$1: Duplicate field 'file'
 at [Source: (byte[])"{"file":{"line":647},"function":"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).watchRuntimeComponents","file":{"name":"coordinator/coordinator.go"}}"; line: 1, column: 155]
	at org.elasticsearch.server@9.0.0/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:706)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:415)
	... 22 more
Caused by: com.fasterxml.jackson.core.JsonParseException: Duplicate field 'file'
 at [Source: (byte[])"{"file":{"line":647},"function":"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).watchRuntimeComponents","file":{"name":"coordinator/coordinator.go"}}"; line: 1, column: 155]
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.JsonReadContext._checkDup(JsonReadContext.java:250)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.JsonReadContext.setCurrentName(JsonReadContext.java:244)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:758)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.JsonGenerator._copyCurrentContents(JsonGenerator.java:2657)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.JsonGenerator.copyCurrentStructure(JsonGenerator.java:2638)
	at org.elasticsearch.xcontent.impl@9.0.0/org.elasticsearch.xcontent.provider.json.JsonXContentGenerator.copyCurrentStructure(JsonXContentGenerator.java:540)
	at org.elasticsearch.xcontent@9.0.0/org.elasticsearch.xcontent.XContentBuilder.copyCurrentStructure(XContentBuilder.java:1237)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper$TypeUtils.decodeAndWriteXContent(XContentDataHelper.java:603)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper$TypeUtils$11.decodeAndWrite(XContentDataHelper.java:502)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper.decodeAndWrite(XContentDataHelper.java:96)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.XContentDataHelper.writeMerged(XContentDataHelper.java:129)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$FieldWriter$IgnoredSource.writeTo(ObjectMapper.java:1084)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:1004)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$FieldWriter$FieldLoader.writeTo(ObjectMapper.java:1059)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:1004)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(SourceLoader.java:217)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:181)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$LeafWithMetrics.source(SourceLoader.java:146)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:290)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:233)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:162)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:72)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:191)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:81)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:754)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:699)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:559)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:79)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:76)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:101)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	... 6 more
	Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder
		at org.elasticsearch.xcontent@9.0.0/org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1282)
		at org.elasticsearch.server@9.0.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:180)
		... 20 more
	Caused by: java.io.IOException: Unclosed object or array found
		at org.elasticsearch.xcontent.impl@9.0.0/org.elasticsearch.xcontent.provider.json.JsonXContentGenerator.close(JsonXContentGenerator.java:600)
		at org.elasticsearch.xcontent@9.0.0/org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1280)
		... 21 more
```

**Comment by martijnvg** [comment]
Looking at the partial source: 
``` 
at [Source: (byte[])"{"file":{"line":647},"function":"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).watchRuntimeComponents","file":{"name":"coordinator/coordinator.go"}}"; line: 1, column: 155]
```

It looks like the object field `file` is in ignored source and gets added to a xcontentbuilder that already has it.

**Comment by kkrik-es** [comment]
Do we have a repro?

**Comment by salvatore-campagna** [comment]
From the partial content I see

```
{
  "file": {
    "line": 647
  },
  "function": "github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).watchRuntimeComponents",
  "file": {
    "name": "coordinator/coordinator.go"
  }
}
```

This looks like synthetic source not correctly merging `line` and `name` under the same `file`

**Comment by lkts** [comment]
Most likely full name of two filed writers is different but leaf name is the same. As a result they don't deduplicate based on the full name.

**Comment by lkts** [comment]
This stack trace looks to me like the duplicate field is in the _input_ of copy meaning we read it from ignored source like that:
```
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.JsonReadContext._checkDup(JsonReadContext.java:250)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.JsonReadContext.setCurrentName(JsonReadContext.java:244)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:758)
	at com.fasterxml.jackson.core@2.17.2/com.fasterxml.jackson.core.JsonGenerator._copyCurrentContents(JsonGenerator.java:2657)
```

**Comment by lkts** [comment]
Likely repro:
```
PUT my-index
{
  "mappings": {
    "_source": {
      "mode": "synthetic"
    },
    "dynamic": false
  }
}

PUT my-index/_doc/0
{
  "top": {
    "file.name": "name",
    "file.line": 400
  }
}

GET my-index/_doc/0
---------------------------------

"error": {
            "root_cause": [
                {
                    "type": "json_parse_exception",
                    "reason": "Duplicate field 'file'\n at [Source: (byte[])\"{\"file\":{\"name\":\"name\"},\"file\":{\"line\":400}}\"; line: 1, column: 31]"
                }
            ],
            "type": "json_parse_exception",
            "reason": "Duplicate field 'file'\n at [Source: (byte[])\"{\"file\":{\"name\":\"name\"},\"file\":{\"line\":400}}\"; line: 1, column: 31]",
            "suppressed": [
                {
                    "type": "illegal_state_exception",
                    "reason": "Failed to close the XContentBuilder",
                    "caused_by": {
                        "type": "i_o_exception",
                        "reason": "Unclosed object or array found"
                    }
                }
            ]
        },
```

**Comment by lkts** [comment]
Fixed by #113910.
